core/src/main/scala/io/dantb/contentless/ContentModel.scala

```diff
@@ -85,20 +95,60 @@ object Validation:
     final case class EntryBlock(size: Option[Size], linkContentType: Option[LinkContentType])
 
   final case class LinkContentType(allowedContentTypes: Set[String], message: Option[String]) extends Validation
-  final case class RichTextMarks(allowedValues: Set[String])                                  extends Validation
-  final case class RichTextNodeTypes(allowedValues: Set[String])                              extends Validation
-  final case class Size(min: Option[Int], max: Option[Int], message: Option[String])          extends Validation
-  case object Unique                                                                          extends Validation
-
-  // equality defers to underlying regex, to allow custom regex subtypes for convenience without breaking equality
-  sealed abstract case class Regexp(regexp: String) extends Validation:
-    override def equals(obj: Any): Boolean = obj.isInstanceOf[Regexp] && obj.asInstanceOf[Regexp].regexp == regexp
-    override def hashCode(): Int           = regexp.hashCode
-  object ValidUrl extends Regexp(Regexp.validUrlRegexp)
-  object Regexp:
-    val validUrlRegexp = "^(ftp|http|https):\\/\\/(\\w+:{0,1}\\w*@)?(\\S+)(:[0-9]+)?(\\/|\\/([\\w#!:.?+=&%@!\\-\\/]))?$"
 
-    def apply(str: String): Regexp = new Regexp(str) {}
+  final case class RichTextMarks(allowedValues: Set[Mark]) extends Validation:
+    def message: Option[String] =
+      val size = allowedValues.size
+      size match
+        case 0 => None
+        case 1 => s"Only ${allowedValues.head.asString} marks are allowed".some
+        case other =>
+          val (allButOne, last) = allowedValues.map(_.asString).splitAt(size - 1)
+          s"Only ${allButOne.mkString(", ")} and ${last.head} marks are allowed".some
+
+  final case class RichTextNodeTypes(allowedValues: Set[RichTextNodeType]) extends Validation:
+    def message: Option[String] =
+      val size = allowedValues.size
+      size match
+        case 0 => None
+        case 1 => s"Only ${allowedValues.head.asString} nodes are allowed".some
+        case other =>
+          val (allButOne, last) = allowedValues.map(_.asString).splitAt(size - 1)
+          s"Only ${allButOne.map(_.replaceAll("-", " ")).mkString(", ")} and ${last.head} nodes are allowed".some
```

Comment:
Nitpicking: Maybe these two could be factored out into a function accepting `"marks"` or `"nodes"` as argument? It's safe as per the `size` matching, but the use of `head` scares me, easy for it to end up in the wrong place when refactoring.

Reply:
Thanks, I'll do a `.toList` + pattern match and remove the duplication - was just being lazy really!